cpp/oneapi/dal/algo/kmeans/backend/gpu/kernels_csr_impl.hpp

```diff
@@ -0,0 +1,412 @@
+/*******************************************************************************
+* Copyright 2024 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#include "oneapi/dal/backend/primitives/reduction.hpp"
+#include "oneapi/dal/backend/interop/common_dpc.hpp"
+#include "oneapi/dal/backend/interop/error_converter.hpp"
+#include "oneapi/dal/backend/interop/table_conversion.hpp"
+#include "oneapi/dal/backend/primitives/ndarray.hpp"
+#include "oneapi/dal/backend/atomic.hpp"
+
+namespace oneapi::dal::kmeans::backend {
+
+using dal::backend::context_gpu;
+using descriptor_t = detail::descriptor_base<task::clustering>;
+using event_vector = std::vector<sycl::event>;
+
+template <typename Data>
+using local_accessor_rw_t = sycl::local_accessor<Data, 1>;
+
+namespace interop = dal::backend::interop;
+namespace pr = dal::backend::primitives;
+namespace de = dal::detail;
+namespace bk = dal::backend;
+
+template <typename Float>
+sycl::event compute_data_squares(sycl::queue& q,
+                                 const pr::ndview<Float, 1>& values,
+                                 const pr::ndview<std::int64_t, 1>& column_indices,
+                                 const pr::ndview<std::int64_t, 1>& row_offsets,
+                                 pr::ndview<Float, 1>& squares) {
+    ONEDAL_PROFILER_TASK(compute_data_squares, q);
+    return pr::reduce_by_rows(q,
+                              values,
+                              column_indices,
+                              row_offsets,
+                              sparse_indexing::zero_based,
+                              squares,
+                              pr::sum<Float>{},
+                              pr::square<Float>{});
+}
+
+// Temporary function, TODO: replace this call with spgemm call
+// TODO: need to add dimensions integer overflow
```

Comment:
is this pending MKL? lets create follow-up task for any TODOs

Reply:
Yes, it should be replaced with MKL-call when it's ready. I have created the task.